pyvelox/complex.cpp

```diff
@@ -0,0 +1,179 @@
+  std::vector<ElementCounter> children;
+};
+} // namespace
+
+void checkOrAssignType(TypePtr& type, const std::function<TypePtr()>& func) {
+  if (type->kind() == TypeKind::UNKNOWN) {
+    type = func();
+  } else if (!(type->kindEquals(func()))) {
+    throw py::type_error(
+        "Cannot construct type tree, invalid variant for complex type");
+  }
+}
+
+template <TypeKind Kind>
+void setElementInFlatVector(
+    vector_size_t idx,
+    const variant& v,
+    VectorPtr& vector) {
+  using NativeType = typename TypeTraits<Kind>::NativeType;
+  auto asFlat = vector->asFlatVector<NativeType>();
+  asFlat->set(idx, NativeType{v.value<NativeType>()});
+}
+
+// Function responsible to construct the type for the passed variant
+// for a complex vector. It is supposed to run a recursive call with a
+// pre-instantiated TypePtr, the target variant and the counter. The
+// passed variant is checked for its data type, and for any complex
+// type involved, the function is called again. The counter here is used
+// to keep in track of the number of elements inserted and the number of
+// types of elements allowed if a complex vector is involved in the variant.
+void constructType(const variant& v, TypePtr& type, ElementCounter& counter) {
+  ++counter.totalElements;
+
+  if (v.isNull()) {
+    // since the variant is NULL, we can't infer the data type
+    // thus it maybe UNKNOWN or INVALID at this stage
+    // which implies further investigation is required
+    if (v.kind() != TypeKind::UNKNOWN && v.kind() != TypeKind::INVALID &&
+        v.kind() != type->kind()) {
+      throw std::invalid_argument("Variant was of an unexpected kind");
+    }
+    return;
+  } else {
+    // if a Non-Null variant's type is unknown or not one of the valid
+    // types which are supported then the Type tree cannot be constructed
+    if (v.kind() == TypeKind::UNKNOWN || v.kind() == TypeKind::INVALID) {
+      throw std::invalid_argument(
+          "Non-null variant has unknown or invalid kind");
+    }
+
+    switch (v.kind()) {
+      case TypeKind::ARRAY: {
+        counter.children.resize(1);
+        auto asArray = v.array();
+        TypePtr childType = createType(TypeKind::UNKNOWN, {});
+        for (const auto& element : asArray) {
+          constructType(element, childType, counter.children[0]);
+        }
+
+        // if child's type still remains Unknown, implies all the
+        // elements in the array are actually NULL
+        if (childType->kind() == TypeKind::UNKNOWN) {
+          throw py::value_error("Cannot construct array with all None values");
+        }
+        checkOrAssignType(type, [&childType]() {
+          return createType<TypeKind::ARRAY>({childType});
+        });
+        break;
+      }
+
+      default: {
+        checkOrAssignType(type, [&v]() { return createScalarType(v.kind()); });
+        break;
+      }
+    }
+  }
+}
+
+// Function is called with the typekind, the variant to be added,
```

Comment:
This comment doesnt seem to match the signature.

Reply:
corrected that, thanks!